### DISCUSSION TITLE
Is it possible to return only the ipv6 address when having both ipv4 and ipv6 addresses

### DISCUSSION THREAD

**Discussion by keeker233** [AdguardTeam/AdGuardHome#4482] [Q&A]
Have a question or an idea? Please search it [on our forum](https://github.com/AdguardTeam/AdGuardHome/discussions) to make sure it was not yet asked. If you cannot find what you had in mind, please [submit it here](https://github.com/AdguardTeam/AdGuardHome/discussions/new).

### Prerequisites

Please answer the following questions for yourself before submitting an issue. **YOU MAY DELETE THE PREREQUISITES SECTION.**

- [x] I am running the latest version
- [x] I checked the documentation and found no answer
- [ ] I checked to make sure that this issue has not already been filed

### Problem Description

In the campus network environment, I hope to use ipv6 and use ipv6 as much as possible, but most software and equipment are still given priority to ipv4 in use. Can I set in adguardhome to only return ipv6 addresses for dual-stack URLs which have both ipv4 address and ipv6 address, so that the device can access it as a pure ipv6 website

### Proposed Solution

Add an option to adguardhome to return only ipv4 or ipv6 addresses if the url has both ipv4 and ipv6 addresses
### Additional Information

At present, I manually set the ipv6 address in dns rewrite, but I cannot easily find the parsed ipv6 address in the query record. Its response is only represented by NOERROR. Is there any convenient way to find the parsed ipv6 address? ipv6 address?

**Comment by fernvenue** [comment]
**AdGuardHome can't and shouldn't do that.**

As we know, DNS Server's response is based on the client's request type:

- Send a request for type `A`, AdGuardHome will and should response with IPv4 address.
- Send a request for type `AAAA`, AdGuardHome will and should response with IPv6 address.

So this is clearly a client behavior, you should set Prefer IPv6 on your device or application. And what AdGuardHome can do is drop all request for a specific type, just like `aaaa_disabled` in `AdGuardHome.yaml` now:

> `aaaa_disabled` — Respond with an empty answer to all AAAA requests. --[AdGuardHome Wiki](https://github.com/AdguardTeam/AdGuardHome/wiki/Configuration#configuration-file)

By the way, it might make sense to change this feature to a specific type that can be set.

**Reply by keeker233** [reply]
Thank you for your suggestion, but now many devices still give priority to ipv4, it should be more convenient if this problem can be solved at the dns server.

**Reply by fernvenue** [reply]
> Thank you for your suggestion, but now many devices still give priority to ipv4, it should be more convenient if this problem can be solved at the dns server.

As I said, AdGuardHome can’t and shouldn’t do that, maybe you can read my answer again :)

**Reply by fernvenue** [reply]
@keeker233 Hi! I would like to add a little clarification, like I said above:

> ...you should set Prefer IPv6 on your device or application.

If you use Windows operating system, you can use `netsh interface ipv6 show prefixpolicies` to make your device prefer IPv6.

**Comment by ainar-g** [comment]
In addition to what @fernvenue said, you can force AdGuard Home to reply with an empty `NOERROR` response to all `A` queries by using the following custom filtering rule:

```none
*$dnstype=A,dnsrewrite=NOERROR
```

Be aware though that using that will disable all `A` queries.  You could also use more precise patterns, e.g. for your campus.  Something like:

```none
||internal.my-campus.edu^$dnstype=A,dnsrewrite=NOERROR
```

**Reply by keeker233** [reply]
Thank you for your answer. I would like to ask what the code dnsrewrite=NOERROR means. Does it tell the client that the query was successful but the IP address was not found? 
In addition, I would like to know why in the query records of adguardhome, some records have only the response code of NOERROR but no specific ip in the response details, while other query records have them. 
I want adguardhome to only return addresses of type aaaa because the bandwidth of ipv6 in my campus network environment is higher than that of ipv4 and will not be limited by traffic, so my main method now is to find some domains in adguard's query records which consumes a lot of traffic and have ipv6 address, and then set it to the corresponding address of the aaaa type in dnsrewrite, but sometimes the specific ip address cannot be found in the records ,does it because the address of the aaaa type is not resolved?
like this 
![image](https://user-images.githubusercontent.com/49800820/163088451-b7ae949e-de05-4fa0-83d2-05752ea3d12e.png)

**Reply by ainar-g** [reply]
@keeker233, sorry for the late response.  It will return a `NODATA` response, which is usually understood as “there may be some information for this domain name, but not the type you asked for”.  It is usually returned when there are e.g. IPv4 addresses but not IPv6 or vice versa.

**Reply by nImEHuntet** [reply]
You are a genius!
We were having issues with our IPT, where a connection to a specific banking website via IPV6 returned as not-reachable.
While the issue was identified, the upstream ISP did little to help the situation.
Your rule was helpful as a workaround because the website works properly via IPv4.

**Comment by 57382** [comment]
> @keeker233, sorry for the late response. It will return a `NODATA` response, which is usually understood as “there may be some information for this domain name, but not the type you asked for”. It is usually returned when there are e.g. IPv4 addresses but not IPv6 or vice versa.



https://github.com/AdguardTeam/AdGuardHome/issues/5919
[https://github.com/AdguardTeam/AdGuardHome/issues/5919]
